Newsletters: add 161 (2021-08-11) (#625)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2021-08-11-newsletter.md

slower when called with the hash of the block that includes the transaction. The review club
evaluated the cause of this performance issue by comparing the steps to retrieve a transaction with
and without txindex.

{% include functions/details-list.md

  q0="What are the different ways `GetTransaction` can retrieve a transaction from disk?"
  a0="The transaction can be retrieved from the mempool (if unconfirmed), by retrieving the entire
  block from disk and searching for the transaction, or by using txindex to fetch the transaction
  from disk by itself."
  a0link="https://bitcoincore.reviews/22383#l-33"

  q1="Why do you think that performance is worse when the block hash is provided (when txindex is
  enabled)?"
  a1="Participants guessed that the bottleneck was in the deserialization of the block. Another
  process unique to fetching the entire block - albeit less time-consuming - is a linear search
  through the entire list of transactions."
  a1link="https://bitcoincore.reviews/22383#l-42"

  q2="If we are looking up the transaction by block hash, what are the steps? How much data
  is deserialized?"

>   a2="We first use the block index to find out which file and byte offset to find the transaction.

Author: Should this be `s/the transaction/the block/`?

Reviewer: Ooh yes, good catch!

projects.  Please consider upgrading to new releases or helping to test
release candidates.*

- [C-Lightning 0.10.1][] is a release
  that contains a number of new features, several bug fixes,
  and a few updates to developing protocols (including [dual
  funding][topic dual funding] and [offers][topic offers]).

- [Bitcoin Core 22.0rc2][bitcoin core 22.0] is a release candidate
  for the next major version of this full node implementation and its
  associated wallet and other software. Major changes in this new
  version include support for [I2P][topic anonymity networks] connections,
  removal of support for [version 2 Tor][topic anonymity networks] connections,
  and enhanced support for hardware wallets.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo],
[C-Lightning][c-lightning repo], [Eclair][eclair repo], [LND][lnd repo],
[Rust-Lightning][rust-lightning repo], [libsecp256k1][libsecp256k1
repo], [Hardware Wallet Interface (HWI)][hwi repo],
[Rust Bitcoin][rust bitcoin repo], [BTCPay Server][btcpay server repo],
[Bitcoin Improvement Proposals (BIPs)][bips repo], and [Lightning
BOLTs][bolts repo].*

> - [Bitcoin Core #21528][] aims to improve the propagation of addresses

Reviewer 2: Given that "addresses" is mostly used in the context of defining recipient scriptPubKeys, it would maybe be good to qualify what sort of addresses we are talking about here. ("node" doesn't feel amazing here, but I haven't been able to think of something better yet.)

```suggestion
- [Bitcoin Core #21528][] aims to improve the propagation of node addresses
```

  in the p2p network. Exposure to a diverse set of addresses is important
  for nodes to be protected against network partitions such as [eclipse attacks][topic eclipse attacks].
  When Bitcoin Core nodes receive an address message containing 10 or fewer
  addresses, they forward it to 1 or 2 peers. This is the primary technique
  used to self-advertise addresses, so sending to peers that would not relay
  these addresses would effectively stop or "black hole" the propagation through

>   the network. Although this cannot be prevented in the malicious case, this

Reviewer 2: ```suggestion
  the network. Although failure to propagate cannot be prevented in the malicious case, this
```

>   patch improves it for the honest cases, such as for block-relay-only

Reviewer 2: ```suggestion
  patch improves address propagation for the honest cases, such as for block-relay-only
```

  connections or light clients.

  This update identifies whether or not an inbound connection is a
  candidate for forwarding addresses based on whether it has sent an address
  related message over the connection, such as `addr`, `addrv2`, or `getaddr`. This behavior
  change could be problematic if there is software on the network that
  relies on receiving address messages but never initiates an address-related
  message. Therefore, the author took care to circulate this proposed change
  before it was merged, including posting it to the
  [mailing list][addrRelay improvements] and researching
  [other open source clients][addr client research] to confirm compatibility.

> - [LND #5484][] allows it to store all data in a single external `etcd`

Reviewer 3: Etcd is the name of a program, so I don't think it should be styled as code.

```suggestion
- [LND #5484][] allows it to store all data in a single external etcd
```

> - [LND #5484][] allows storing all data in a single external Etcd

Reviewer 3: etcd is styled in lowercase:

```suggestion
- [LND #5484][] allows storing all data in a single external etcd
```

or we could make this a link:

```suggestion
- [LND #5484][] allows storing all data in a single external [etcd](https://etcd.io/)
```

Author: Happy to discuss this further to find a consensus opinion, but my own personal preference is that we treat program names as either proper nouns (i.e. capitalized) or as code (e.g. in backtics).  (Up to the author of a section which they want to use; they may even vary within the section as needed, e.g. with "LND" and "`lnd'".)

Reviewer 3: It's not hugely important, but I'm not sure that we should do this. If a program has a widely used lower-case style, then how do we decide how to capitalize it? Why 'LND' instead of 'Lnd'? Why 'Etcd' instead of 'EtcD' (etcd stands for [etc distributed](https://dbdb.io/db/etcd))? Would we capitalize glibc as Glibc or GLibC?

Like I said, it's not really that important, so I'm happy to defer to your choice.

Author: In those cases, I think we do the same thing we already do for Bitcoin Core and C-Lightning: when referring to the project, we use the full name (Bitcoin Core, C-Lightning, Etc Distributed, GNU Lib C); when referring to a specific program or library, put it in backtics (`bitcoind`, `lightningd`, `etcd`, `glibc`).

My overall thought here is that I really like to make text as easy and fast to read as possible, and I think conventions like starting proper nouns with a capital letter contribute to that.

>   database. This improves high-availability deployments by making cluster

Reviewer 3: 'it' refers to LND, which hasn't been referred to before in the sentence, so this is a little bit awkward. Perhaps "LND #5484 allows the node to store all data in a single external `etcd` database." or "LND #5484 allows all of the node's data to be stored in a single external `etcd` database."

  leadership changes instantaneous. The corresponding LND clustering

>   documentation was previously covered [here][news157 lnd ha].

Reviewer 2: ```suggestion
  documentation was previously covered in [Newsletter #157][news157 lnd ha].
```

> - [Rust-Lightning #1004][] Adds a new event for `PaymentForwarded` that

Reviewer 3: ```suggestion
- [Rust-Lightning #1004][] adds a new event for `PaymentForwarded` that
```

  allows tracking when a payment has been successfully forwarded.  Since
  successful forwarding may earn fees for the node, this allows tracking
  that income for the user's accounting records.

- [BTCPay Server #2730][] makes the amount optional when generating
  invoices. This simplifies the payment flow in cases where the operator
  delegates the choice of the amount to the user, e.g. when topping up
  an account.

{% include references.md %}
{% include linkers/issues.md issues="21528,5484,1004,2730,22383" %}

> [C-Lightning 0.10.1]: https://github.com/ElementsProject/lightning/releases/tag/v0.10.1rc2

Reviewer 4: The link points to the rc version of 0.10.1

Author: Fixed, thanks @shesek!
